ecosystem/gstreamer_plugin/gst_mtl_st40_rx.c

```diff
@@ -0,0 +1,509 @@
+  gint ret;
+  void* mbuf = NULL;
+
+  clock_gettime(CLOCK_REALTIME, &ts);
+  ts.tv_sec += src->timeout_mbuf_get_seconds;
+  ret = pthread_mutex_timedlock(&(src->get_mbuff_mutex), &ts);
+
+  if (!ret)
+    ret = pthread_cond_timedwait(&(src->get_mbuff_cond), &(src->get_mbuff_mutex), &ts);
+
+  if (ret == ETIMEDOUT) return NULL;
+
+  pthread_mutex_unlock(&(src->get_mbuff_mutex));
+
+  mbuf = st40_rx_get_mbuf(src->rx_handle, usrptr, size);
+  if (!mbuf) GST_ERROR("Failed to get ancillary mbuf\n");
+
+  return mbuf;
+}
+
+static GstFlowReturn gst_mtl_st40_rx_fill_buffer(Gst_Mtl_St40_Rx* src, GstBuffer** buffer,
+                                                 void* usrptr) {
+  struct st40_rfc8331_rtp_hdr* hdr;
+  struct st40_rfc8331_payload_hdr* payload_hdr;
+  GstMapInfo dest_info;
+  guint16 data, fill_size;
+  gint udw_size;
+
+  hdr = (struct st40_rfc8331_rtp_hdr*)usrptr;
+  payload_hdr = (struct st40_rfc8331_payload_hdr*)(&hdr[1]);
+  udw_size = payload_hdr->second_hdr_chunk.data_count & 0xff;
+
+  if (src->udw_size == 0) {
+    src->udw_size = udw_size;
+    src->anc_data = (char*)malloc(udw_size);
+  } else if (src->udw_size != udw_size) {
+    GST_WARNING("Size of recieved ancillary data has changed");
+    if (src->anc_data) {
+      free(src->anc_data);
+      src->anc_data = NULL;
+    }
+    src->udw_size = udw_size;
+    src->anc_data = (char*)malloc(udw_size);
+  }
+
+  if (udw_size == 0) {
+    GST_ERROR("Ancillary data size is 0");
+    return GST_FLOW_ERROR;
+  }
+
+  *buffer = gst_buffer_new_allocate(NULL, src->udw_size, NULL);
+  if (!*buffer) {
+    GST_ERROR("Failed to allocate space for the buffer");
+    return GST_FLOW_ERROR;
+  }
+
+  if (!gst_buffer_map(*buffer, &dest_info, GST_MAP_WRITE)) {
+    GST_ERROR("Failed to map the buffer");
+    return GST_FLOW_ERROR;
+  }
+
+  for (int i = 0; i < udw_size; i++) {
+    data = st40_get_udw(i + 3, (uint8_t*)&payload_hdr->second_hdr_chunk);
+    if (!st40_check_parity_bits(data)) {
+      GST_ERROR("Ancillary data parity bits check failed");
+      return GST_FLOW_ERROR;
+    }
+    src->anc_data[i] = data & 0xff;
+  }
+
+  fill_size = gst_buffer_fill(*buffer, 0, src->anc_data, udw_size);
+  gst_buffer_unmap(*buffer, &dest_info);
+
+  if (fill_size != src->udw_size) {
+    GST_WARNING("Failed to fill buffer");
+  }
+
+  return GST_FLOW_OK;
+}
+
+static GstFlowReturn gst_mtl_st40_rx_create(GstBaseSrc* basesrc, guint64 offset,
+                                            guint length, GstBuffer** buffer) {
+  Gst_Mtl_St40_Rx* src = GST_MTL_ST40_RX(basesrc);
+  void *mbuf, *usrptr;
+  guint16 size;
+  gint ret;
+
+  GST_OBJECT_LOCK(src);
+
+  /* get the mbuff */
+  mbuf = st40_rx_get_mbuf(src->rx_handle, &usrptr, &size);
+  if (!mbuf) {
+    mbuf = gst_mtl_st40_rx_get_mbuf_with_timeout(src, src->rx_handle, &usrptr, &size);
+  }
+
+  if (size == 0) {
+    GST_ERROR("No anciallry data recived");
+    GST_OBJECT_UNLOCK(src);
+    return GST_FLOW_ERROR;
+  }
+
+  if (!mbuf) {
```

Comment:
here you interpret timeout as end of stream, what if other error occurred instead of timeout?

what will be the value of `size` if we get timeout? it is not assigned anywhere so it is random, can be 0, so we can return with GST_FLOW_ERROR on timeout.

I don't like this return values handling.

Reply:
Yeaaa the size is moved after the timeout handling 

We kinda have no way of knowing if any error occurred and if we can only work with what we have if we don't have the mbuff for the plugin it's the EOS 

if there were errors ? 
Still EOS the mtl said its ready unlocked the mutex and still return mbuff which means we don't have anything to work with 

Ooooor we timed out 
Basicly the result is the same we just say ok I'm out